### PR TITLE
Test for transposing a Square SGSarseMatrix.

### PR DESCRIPTION
**Pull request by sudk1896**
Adding a test for transposing a square `SGSparseMatrix`.
#2748 
@karlnapf Review, I implemented your style fixes (remove camel case in my new test).

### CODE REVIEW

**Comment by sudk1896** [comment]
@karlnapf : The valgrind output before adding the test is -
```
sudarshan@sudarshan-HP-15-Notebook-PC:~/shogun/build/tests/unit$ valgrind --leak-check=full --show-leak-kinds=all ./shogun-unit-test --gtest_filter=*SGSparseMatrix*
==4082== Memcheck, a memory error detector
==4082== Copyright (C) 2002-2013, and GNU GPL'd, by Julian Seward et al.
==4082== Using Valgrind-3.10.1 and LibVEX; rerun with -h for copyright info
==4082== Command: ./shogun-unit-test --gtest_filter=*SGSparseMatrix*
==4082== 
Note: Google Test filter = *SGSparseMatrix*
[==========] Running 10 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 10 tests from SGSparseMatrix
[ RUN      ] SGSparseMatrix.multiply_float64_int32
[       OK ] SGSparseMatrix.multiply_float64_int32 (23 ms)
[ RUN      ] SGSparseMatrix.multiply_complex128_int32
[       OK ] SGSparseMatrix.multiply_complex128_int32 (13 ms)
[ RUN      ] SGSparseMatrix.multiply_complex128_float64
[       OK ] SGSparseMatrix.multiply_complex128_float64 (10 ms)
[ RUN      ] SGSparseMatrix.access_by_index
[       OK ] SGSparseMatrix.access_by_index (9 ms)
[ RUN      ] SGSparseMatrix.io_libsvm
[       OK ] SGSparseMatrix.io_libsvm (217 ms)
[ RUN      ] SGSparseMatrix.access_by_index_non_square
[       OK ] SGSparseMatrix.access_by_index_non_square (37 ms)
[ RUN      ] SGSparseMatrix.get_transposed_more_features
[       OK ] SGSparseMatrix.get_transposed_more_features (50 ms)
[ RUN      ] SGSparseMatrix.get_transposed_more_vectors
[       OK ] SGSparseMatrix.get_transposed_more_vectors (40 ms)
[ RUN      ] SGSparseMatrix.from_dense
[       OK ] SGSparseMatrix.from_dense (31 ms)
[ RUN      ] SGSparseMatrix.test_transposed_square_matrix
[       OK ] SGSparseMatrix.test_transposed_square_matrix (82 ms)
[----------] 10 tests from SGSparseMatrix (519 ms total)

[----------] Global test environment tear-down
[==========] 10 tests from 1 test case ran. (568 ms total)
[  PASSED  ] 10 tests.
==4082== 
==4082== HEAP SUMMARY:
==4082==     in use at exit: 104 bytes in 1 blocks
==4082==   total heap usage: 115,379 allocs, 115,378 frees, 57,768,018 bytes allocated
==4082== 
==4082== 104 bytes in 1 blocks are still reachable in loss record 1 of 1
==4082==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==4082==    by 0x740B7A8: ??? (in /usr/lib/x86_64-linux-gnu/libgomp.so.1.0.0)
==4082==    by 0x740FD8A: ??? (in /usr/lib/x86_64-linux-gnu/libgomp.so.1.0.0)
==4082==    by 0x740B9CC: omp_set_dynamic (in /usr/lib/x86_64-linux-gnu/libgomp.so.1.0.0)
==4082==    by 0x56C9774: shogun::Parallel::Parallel() (in /home/sudarshan/shogun/build/src/shogun/libshogun.so.17.2)
==4082==    by 0x56C95F6: shogun::init_shogun(void (*)(_IO_FILE*, char const*), void (*)(_IO_FILE*, char const*), void (*)(_IO_FILE*, char const*), void (*)(bool&, bool&)) (in /home/sudarshan/shogun/build/src/shogun/libshogun.so.17.2)
==4082==    by 0xCA4231: main (in /home/sudarshan/shogun/build/tests/unit/shogun-unit-test)
==4082== 
==4082== LEAK SUMMARY:
==4082==    definitely lost: 0 bytes in 0 blocks
==4082==    indirectly lost: 0 bytes in 0 blocks
==4082==      possibly lost: 0 bytes in 0 blocks
==4082==    still reachable: 104 bytes in 1 blocks
==4082==         suppressed: 0 bytes in 0 blocks
==4082== 
==4082== For counts of detected and suppressed errors, rerun with: -v
==4082== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

**Comment by sudk1896** [comment]
Valgrind output after adding the test is -
```
sudarshan@sudarshan-HP-15-Notebook-PC:~/shogun/build/tests/unit$ valgrind --leak-check=full --show-leak-kinds=all ./shogun-unit-test --gtest_filter=*SGSparseMatrix*
==4096== Memcheck, a memory error detector
==4096== Copyright (C) 2002-2013, and GNU GPL'd, by Julian Seward et al.
==4096== Using Valgrind-3.10.1 and LibVEX; rerun with -h for copyright info
==4096== Command: ./shogun-unit-test --gtest_filter=*SGSparseMatrix*
==4096== 
Note: Google Test filter = *SGSparseMatrix*
[==========] Running 10 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 10 tests from SGSparseMatrix
[ RUN      ] SGSparseMatrix.multiply_float64_int32
[       OK ] SGSparseMatrix.multiply_float64_int32 (22 ms)
[ RUN      ] SGSparseMatrix.multiply_complex128_int32
[       OK ] SGSparseMatrix.multiply_complex128_int32 (14 ms)
[ RUN      ] SGSparseMatrix.multiply_complex128_float64
[       OK ] SGSparseMatrix.multiply_complex128_float64 (10 ms)
[ RUN      ] SGSparseMatrix.access_by_index
[       OK ] SGSparseMatrix.access_by_index (9 ms)
[ RUN      ] SGSparseMatrix.io_libsvm
[       OK ] SGSparseMatrix.io_libsvm (218 ms)
[ RUN      ] SGSparseMatrix.access_by_index_non_square
[       OK ] SGSparseMatrix.access_by_index_non_square (35 ms)
[ RUN      ] SGSparseMatrix.get_transposed_more_features
[       OK ] SGSparseMatrix.get_transposed_more_features (46 ms)
[ RUN      ] SGSparseMatrix.get_transposed_more_vectors
[       OK ] SGSparseMatrix.get_transposed_more_vectors (39 ms)
[ RUN      ] SGSparseMatrix.from_dense
[       OK ] SGSparseMatrix.from_dense (29 ms)
[ RUN      ] SGSparseMatrix.test_transposed_square_matrix
[       OK ] SGSparseMatrix.test_transposed_square_matrix (82 ms)
[----------] 10 tests from SGSparseMatrix (513 ms total)

[----------] Global test environment tear-down
[==========] 10 tests from 1 test case ran. (563 ms total)
[  PASSED  ] 10 tests.
==4096== 
==4096== HEAP SUMMARY:
==4096==     in use at exit: 104 bytes in 1 blocks
==4096==   total heap usage: 115,379 allocs, 115,378 frees, 57,768,018 bytes allocated
==4096== 
==4096== 104 bytes in 1 blocks are still reachable in loss record 1 of 1
==4096==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==4096==    by 0x740B7A8: ??? (in /usr/lib/x86_64-linux-gnu/libgomp.so.1.0.0)
==4096==    by 0x740FD8A: ??? (in /usr/lib/x86_64-linux-gnu/libgomp.so.1.0.0)
==4096==    by 0x740B9CC: omp_set_dynamic (in /usr/lib/x86_64-linux-gnu/libgomp.so.1.0.0)
==4096==    by 0x56C9774: shogun::Parallel::Parallel() (in /home/sudarshan/shogun/build/src/shogun/libshogun.so.17.2)
==4096==    by 0x56C95F6: shogun::init_shogun(void (*)(_IO_FILE*, char const*), void (*)(_IO_FILE*, char const*), void (*)(_IO_FILE*, char const*), void (*)(bool&, bool&)) (in /home/sudarshan/shogun/build/src/shogun/libshogun.so.17.2)
==4096==    by 0xCA4231: main (in /home/sudarshan/shogun/build/tests/unit/shogun-unit-test)
==4096== 
==4096== LEAK SUMMARY:
==4096==    definitely lost: 0 bytes in 0 blocks
==4096==    indirectly lost: 0 bytes in 0 blocks
==4096==      possibly lost: 0 bytes in 0 blocks
==4096==    still reachable: 104 bytes in 1 blocks
==4096==         suppressed: 0 bytes in 0 blocks
==4096== 
==4096== For counts of detected and suppressed errors, rerun with: -v
==4096== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
sudarshan@sudarshan-HP-15-Notebook-PC:~/shogun/build/tests/unit$ 
```

**Comment by sudk1896** [comment]
@karlnapf : The test hasn't introduced any new memory leak as you can see.

File: tests/unit/lib/SGSparseMatrix_unittest.cc [hunk 1]
```diff
@@ -265,3 +265,25 @@ TEST(SGSparseMatrix, from_dense)
...
+	GenerateMatrix<SGSparseMatrix<float64_t> >(sparse_level, numberOfFeatures, numberOfVectors, randSeed, &sparseMatrix);
+
+	SGSparseMatrix<float64_t> sparseMatrixT=sparseMatrix.get_transposed();
+
+	//check dimension
+	EXPECT_EQ(sparseMatrix.num_features, sparseMatrixT.num_vectors);
+	EXPECT_EQ(sparseMatrix.num_vectors, sparseMatrixT.num_features);
+
+	//check contents
+	for (index_t featIndex=0; featIndex<numberOfFeatures; ++featIndex)
```
Comment (lambday) [inline tests/unit/lib/SGSparseMatrix_unittest.cc]: Hey @sudk1896, I still see `CamelCases` in variable naming here. So I think there has been a confusion in understanding what @karlnapf was saying about styling issue. What he meant was, if I am not wrong, to send your new test with the style we expect our code to be in. This patch is for adding a *new test* - so let's make that the purpose of this patch. Style change for *existing code* is another issue. Let's make a different patch for that purpose.

Doesn't make sense to have a patch for *new code addition* in a style that we don't want, only to have the style reverted in a subsequent patch.

Please have a look at our [coding style](https://github.com/shogun-toolbox/shogun/wiki/Code-style) wiki.

Comment (sudk1896) [inline tests/unit/lib/SGSparseMatrix_unittest.cc]: @lambday: Thanks. Now I get it. So I only change the `CamelCase` in my newly added test and not in the other (older) tests. That is what you meant, correct ?

Comment (lambday) [inline tests/unit/lib/SGSparseMatrix_unittest.cc]: Please have a look at `NAMING CONVENTIONS:` section in the above wiki.

**Comment by sudk1896** [comment]
@lamday: I added the braces to the two for loops. I don' t think it enhances readability but the style guide does mention it though.

**Comment by sudk1896** [comment]
@lambday: I added the braces in the for loops at the end. I don't think it enhances readability but the style guide mentions it. If this is ready to merge, let me know, I will squash all commits into one then.

**Comment by lambday** [comment]
@sudk1896 `I don't think it enhances readability but the style guide mentions it.` - I don't think it mentions that :)

File: tests/unit/lib/SGSparseMatrix_unittest.cc [hunk 2]
```diff
@@ -265,3 +265,27 @@ TEST(SGSparseMatrix, from_dense)
...
+
+	SGSparseMatrix<float64_t> sparse_matrix_t=sparse_matrix.get_transposed();
+
+	//check dimension
+	EXPECT_EQ(sparse_matrix.num_features, sparse_matrix_t.num_vectors);
+	EXPECT_EQ(sparse_matrix.num_vectors, sparse_matrix_t.num_features);
+
+	//check contents
+	for (index_t feat_index=0; feat_index<number_of_features; ++feat_index){
+		for (index_t vec_index=0; vec_index<number_of_vectors; ++vec_index){
```
Comment (lambday) [inline tests/unit/lib/SGSparseMatrix_unittest.cc]: No braces for single **line** loops/conditionals. Check styling guide.

File: tests/unit/lib/SGSparseMatrix_unittest.cc [hunk 3]
```diff
@@ -265,3 +265,27 @@ TEST(SGSparseMatrix, from_dense)
...
+	GenerateMatrix<SGSparseMatrix<float64_t> >(sparse_level, number_of_features, number_of_vectors, rand_seed, &sparse_matrix);
+
+	SGSparseMatrix<float64_t> sparse_matrix_t=sparse_matrix.get_transposed();
+
+	//check dimension
+	EXPECT_EQ(sparse_matrix.num_features, sparse_matrix_t.num_vectors);
+	EXPECT_EQ(sparse_matrix.num_vectors, sparse_matrix_t.num_features);
+
+	//check contents
+	for (index_t feat_index=0; feat_index<number_of_features; ++feat_index){
```
Comment (lambday) [inline tests/unit/lib/SGSparseMatrix_unittest.cc]: `{` on a new line. Check styling guide.

**Comment by karlnapf** [comment]
Guys can we please stop spamming this thread with this overly verbose style discussion? I very explicitly and clearly said how it has to be in my first response and then repeated it twice.  Shogun code is public to get inspiration and google is full of tutorials how to send clean patches that do nix mix up whitespace or style changes with newly added code.

**Comment by karlnapf** [comment]
We can merge this once it is squashed.

BTW for the future, use `git commit --amend`  -- it avoids having all these commits and the need for squashing for smaller patches -- google its usage

**Comment by karlnapf** [comment]
updates here?

**Pull request closed without merging**